### PR TITLE
Rest api infra

### PR DESCRIPTION
**Pull request by dsschult**
Basic rest api infrastructure.

### CODE REVIEW

File: lta/config.py [hunk 1]
```diff
@@ -2,15 +2,19 @@
 
 import os
 
+NA = (0,) # use a tuple so 'is' operator works correctly
```
Comment (blinkdog) [inline lta/config.py]: I didn't know if this should be renamed for better external usage:

```Python 3
my_config = {
    "REST_HOST": "localhost",
    "REST_PORT": config.ENV_REQUIRED
}
my_config = config.from_environment(my_config)
```

But it seems like overkill. `NA` works well enough.

File: lta/config.py [hunk 2]
```diff
@@ -2,15 +2,19 @@
...
 
 def from_environment(keys):
```
Comment (blinkdog) [inline lta/config.py]: I was really impressed that no additional unit tests were required to keep 100% coverage and add support for dictionary input.

File: setupenv.sh [hunk 3]
```diff
@@ -3,4 +3,4 @@ unset PYTHONPATH
 virtualenv -p python3 env
 echo "unset PYTHONPATH" >> env/bin/activate
 . env/bin/activate
-pip install requests tornado git+https://github.com/WIPACrepo/rest-tools
+pip install requests tornado pytest pytest-mock pytest-cov git+https://github.com/WIPACrepo/rest-tools
```
Comment (blinkdog) [inline setupenv.sh]: Sorry, I haven't been updating `setupenv.sh` very well.
I'll make it a point to update this file if I add any dependencies to the project.

File: lta/rest_server.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
+"""
+Long Term Archive REST API server.
+
+Run with `python -m lta.rest_server`.
+"""
+
+import asyncio
```
Comment (blinkdog) [inline lta/rest_server.py]:
```suggestion
```

File: lta/rest_server.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+import logging
+
+from tornado.ioloop import IOLoop
+
+from rest_tools.client import json_decode
```
Comment (blinkdog) [inline lta/rest_server.py]:
```suggestion
```

File: lta/rest_server.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+from rest_tools.server import RestServer, RestHandler, RestHandlerSetup, authenticated, catch_error
+
```
Comment (blinkdog) [inline lta/rest_server.py]:
```suggestion
import tornado
from functools import wraps

```

File: lta/rest_server.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        async def wrapper(self, *args, **kwargs):
+            roles = _auth.get('roles', [])
+
+            authorized = False
+
+            auth_role = self.auth_data.get('long-term-archive',{}).get('role',None)
+            if roles and auth_role in roles:
+                authorized = True
+            else:
+                logging.info('roles: %r    attrs: %r', roles, attrs)
```
Comment (blinkdog) [inline lta/rest_server.py]: Where does this `attrs` come from?

File: lta/rest_server.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        super(BaseLTAHandler, self).initialize(*args, **kwargs)
+        self.db = db
+
+class MainHandler(BaseLTAHandler):
+    def get(self):
+        self.write({})
+
+def start(debug=False):
+    config = from_environment(EXPECTED_CONFIG)
+    logger = logging.getLogger('lta.rest')
```
Comment (blinkdog) [inline lta/rest_server.py]:
```suggestion
```

File: lta/rest_server.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    args['db'] = {} # this could be a DB, but a dict works for now
+
+    server = RestServer(debug=debug)
+    server.add_route('/', MainHandler, args)
+    server.startup(address=config['LTA_REST_URL'],
+                   port=config['LTA_REST_PORT'])
+    return server
+
+if __name__ == '__main__':
+    logging.basicConfig(level=logging.DEBUG)
```
Comment (blinkdog) [inline lta/rest_server.py]: After this PR is merged, can you open an issue for me to refactor this to structured logging?

File: lta/rest_server.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    is not necessary, as this decorator will perform authentication
+    checking as well.
+
+    Args:
+        roles (list): The roles to match
+
+    Raises:
+        :py:class:`tornado.web.HTTPError`
+    """
+    def make_wrapper(method):
```
Comment (blinkdog) [inline lta/rest_server.py]: Missing coverage tests

File: lta/rest_server.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+                logging.info('role mismatch')
+
+            if not authorized:
+                raise tornado.web.HTTPError(403, reason="authorization failed")
+
+            return await method(self, *args, **kwargs)
+        return wrapper
+    return make_wrapper
+
+class BaseLTAHandler(RestHandler):
```
Comment (blinkdog) [inline lta/rest_server.py]: Missing coverage tests

File: lta/rest_server.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    def initialize(self, db, *args, **kwargs):
+        super(BaseLTAHandler, self).initialize(*args, **kwargs)
+        self.db = db
+
+class MainHandler(BaseLTAHandler):
```
Comment (blinkdog) [inline lta/rest_server.py]: Missing coverage tests

File: tests/test_rest_server.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+from rest_tools.client import AsyncSession
+
+def test_server_startup():
+    """
+    Check that there are no errors during server startup.
+    """
+    s = start()
+    s.stop()
+
+async def test_server_reachability(monkeypatch):
```
Comment (blinkdog) [inline tests/test_rest_server.py]:
```suggestion
@pytest.mark.asyncio
async def test_server_reachability(monkeypatch):
```

File: lta/config.py (hunk 1 above)
Comment (dsschult) [reply lta/config.py]: Resolving this by making NA private.

File: lta/rest_server.py (hunk 5 above)
Comment (dsschult) [reply lta/rest_server.py]: I'm going to need this shortly, so the import stays.

File: lta/rest_server.py (hunk 9 above)
Comment (dsschult) [reply lta/rest_server.py]: We talked about this, and this is just a standin for now to get logging info.  Will be revised at a later time.

File: lta/rest_server.py (hunk 10 above)
Comment (dsschult) [reply lta/rest_server.py]: It's also not used by any code yet.  Will come with the next pull request.

File: lta/rest_server.py (hunk 11 above)
Comment (dsschult) [reply lta/rest_server.py]: I'm unsure why this, and the following, aren't getting covered. They should be, and the test is returning the right answer.  Maybe a bug in the coverage settings.

Comment (dsschult) [reply lta/rest_server.py]: Might be this:
```
tests/test_rest_server.py::test_server_reachability
  /home/dschultz/Documents/github/lta/env/lib/python3.7/site-packages/_pytest/python.py:165: RuntimeWarning: coroutine 'test_server_reachability' was never awaited
    testfunction(**testargs)
```

Comment (dsschult) [reply lta/rest_server.py]: solved with latest commit

File: lta/rest_server.py (hunk 6 above)
Comment (blinkdog) [reply lta/rest_server.py]: You need this `import tornado` for Line 56

Comment (dsschult) [reply lta/rest_server.py]: ah, I see.  actually, I need `import tornado.web`

**Pull request merged**